phenol-annotations/src/main/java/org/monarchinitiative/phenol/annotations/obo/go/GoGeneAnnotationParser.java

```diff
@@ -35,104 +37,57 @@
-   */
-  public GoGeneAnnotationParser(File file) throws PhenolException {
-    this.file = file;
-    try {
-      this.reader = new BufferedReader(new FileReader(file));
-      this.nextLine = reader.readLine();
-      skipHeaderAndCheckFirst();
-      ImmutableList.Builder<GoGaf21Annotation> builder = new ImmutableList.Builder<>();
-      while (hasNext()) {
-        GoGaf21Annotation annot = next();
-        builder.add(annot);
-      }
-      reader.close();
-      annotations=builder.build();
-    } catch (IOException | TermAnnotationParserException e) {
-      String msg=String.format("Could not parse GO annotation file: %s",e.getMessage());
-      throw new PhenolException(msg);
-    }
+  private GoGeneAnnotationParser() {
   }
 
-  public GoGeneAnnotationParser(String absolutepath)throws PhenolException {
-    this(new File(absolutepath));
-  }
 
-  /**
-   * Skip all header lines and check next data line.
-   *
-   * @throws TermAnnotationParserException If the first line is not as expected.
-   * @throws IOException If there is a problem with reading from the file.
-   */
-  private void skipHeaderAndCheckFirst() throws TermAnnotationParserException, IOException {
-    skipUntilData();
-    if (nextLine == null) {
-      throw new TermAnnotationParserException("GAF2.1 file contained no data!");
-    }
-    final String[] arr = nextLine.split("\t");
-    if (arr.length < 15 || arr.length > 17) {
-      throw new TermAnnotationParserException(
-          "First line of file had "
-              + arr.length
-              + " columns, but expected between 15 and 17 entries.");
-    }
+  public static List<GoGaf21Annotation> loadAnnotations(String filename) {
+    File f = new File(filename);
+    return GoGeneAnnotationParser.loadAnnotations(f);
   }
 
-  /*
-   * Skip comment lines until data
-   *
-   */
-
-  private void skipUntilData() throws IOException {
-    while (nextLine == null || nextLine.startsWith("!")) {
-      nextLine = reader.readLine();
+  public static List<GoGaf21Annotation> loadAnnotations(File file) {
+    String nextLine;
+    ImmutableList.Builder<GoGaf21Annotation> builder = new ImmutableList.Builder<>();
+    try (BufferedReader br = new  BufferedReader(new FileReader(file))) {
+      nextLine =br.readLine();
+      // skip the comments, which start with exclamation marks
+      while (nextLine.isEmpty() || nextLine.startsWith("!")) {
+        nextLine = br.readLine();
+      }
+      do {
+        final String[] arry = nextLine.split("\t");
+        try {
+          GoGaf21Annotation annot = new GoGaf21Annotation(arry);
+          builder.add(annot);
+        } catch (PhenolException e) {
+          System.err.printf("[ERROR] Could not parse GoGaf line (%s): %s", nextLine, e.getMessage());
+          // just skip this error. Should actually never happen
+        }
+      } while ((nextLine = br.readLine()) != null);
```

Comment:
you're calling `nextLine = br.readLine()` a three times in the `try` block. Why not just call it once? 

```java
try (BufferedReader br = new  BufferedReader(new FileReader(file))) {
    while ((nextLine = br.readLine()) != null) {
        // skip the comments, which start with exclamation marks
        if (nextLine.isEmpty() || nextLine.startsWith("!")) {
            continue;         
        } else {
            final String[] arry = nextLine.split("\t");
            try {
                GoGaf21Annotation annot = new GoGaf21Annotation(arry);
                builder.add(annot);
...

Reply:
Thanks for this tidbit of elegant code, Master Jules, much appreciated :-0 !!

Reply:
(bow)